FixedBed0D Updates Scaling and Initialization

bpaul4 opened the pull request:
## Fixes
This completes internal scaling for the FB0D model based on recent updates to the MB and BFB models.

## Summary/Motivation:


## Changes proposed in this PR:
- Expanded model scaling for constraints and variables
- New scaling tests
- Minor initialization changes

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

andrewlee94 commented on idaes/models_extra/gas_solid_contactors/unit_models/fixed_bed_0D.py:
```diff
@@ -26,6 +30,7 @@
     UnitModelBlockData,
 )
 from idaes.core.util.config import is_physical_parameter_block
+from idaes.core.util.exceptions import InitializationError
```
I do not see this being used later in the new code.

bpaul4 replied:
Ah, thanks - I'll clean up the imports. I had attempted to implement an `InitializationError` check similar to the `PressureChanger` initialization on [lines 759-777](https://github.com/IDAES/idaes-pse/blob/a7680abb68c179e7c3abd27ddcbfe5a3f39747c3/idaes/models/unit_models/pressure_changer.py#L759), but found that inserting an extra solve statement in order to check the termination condition spawned other issues. Please let me know if you think the FB0D should use the `InitializationError` or if it is simple enough not to need it.

andrewlee94 replied:
You should have the `InitialziationError`. However, do you need an additional solve to do it? An initialization routine should always finish with ha solve of the full model (unless it happens to have no active constraints), so there should generally be a solution for you to check. The only case where you can skip the check is if you have no active constraints (and even then it should be conditional on there being no active constraints).

bpaul4 replied:
Currently, the initialization routine does not have any full model solves - it initializes the gas/solid thermophysical properties and freezes them, initializes the reactions properties and releases the property states. I would need to add a Step 2 which is a full solve similar to the linked PressureChanger script.

andrewlee94 commented on idaes/models_extra/gas_solid_contactors/unit_models/fixed_bed_0D.py:
```diff
@@ -446,10 +452,97 @@ def initialize(
+        if hasattr(self, "bed_diameter"):
+            if iscale.get_scaling_factor(self.bed_diameter) is None:
+                sf = 1 / value(self.bed_diameter)
+                iscale.set_scaling_factor(self.bed_diameter, 1e-1 * sf)
```
It seems a bit odd that your are multiplying the inverse of the current value by 0.1 to get the scaling factor. A comment on why htis is necessary would probably be useful.

chineduobiora replied:
@bpaul4 it is okay to remove this.

andrewlee94 commented on idaes/models_extra/gas_solid_contactors/unit_models/tests/test_FB0D.py:
```diff
@@ -176,6 +179,179 @@ def test_build(self, iron_oc):
+    def test_initialize_unscaled(self, iron_oc_unscaled):
+        optarg = {
+            "bound_push": 1e-8,
+            "halt_on_ampl_error": "yes",
```
A test should probably not have `halt_on_ampl_error` set to `"yes"`. Similarly, I would not set a specific linear solver here (especially seeing as MA27 is the default anyway).

Finally, it it necessary to set `bound_push` like this? Sometimes it is, but it should be avoided if not necessary (and scaling often resolves many of these issues).

andrewlee94 commented on idaes/models_extra/gas_solid_contactors/unit_models/tests/test_FB0D.py:
```diff
@@ -176,6 +179,179 @@ def test_build(self, iron_oc):
+
+        solver = get_solver("ipopt", optarg)  # create solver
+
+        initialize_by_time_element(iron_oc_unscaled.fs,
```
First, this looks like it should be separate from the `initialization_tester` above.

Secondly, there should be some assert statement following this to ensure that it actually did what was intended.

I think this also applies to some other similar tests later in this file.

andrewlee94 commented on idaes/models_extra/gas_solid_contactors/unit_models/tests/test_FB0D.py:
```diff
@@ -176,6 +179,179 @@ def test_build(self, iron_oc):
+
+        # Check for optimal solution
+        assert check_optimal_termination(results)
+        assert results.solver.status == SolverStatus.ok
```
This is not necessary - the `check_optimal_termination` method handles this.

bpaul4 replied:
See earlier comment on `InitializationError`.

andrewlee94 reviewed (changes requested):
A few general comments, but I think only one of them is critical.

bpaul4 commented:
I've addressed the majority of the review comments, including adding the `InitializationError` back in and separating the `initialize_by_time` into it's own test for each of the four cases (scaled and unscaled, both with and without energy balance).

The only remaining failing test is `test_initialize_unscaled` for the case with an energy balance and it's unclear why constraint 1731 is failing or which equation it is (I added some flags to the unit model and test code so the exact error will appear in the workflow run log, I'll clean them up later). It seems the constraint indexing is not the same between the test and example problems, so I checked every constraint in the model. The only expression that exponentiates variables is the rate constant equation through the temperature variable and I would be surprised if this is causing the issue. I also was not able to reproduce the solver error when modifying the TGA flowsheet example to remove scaling and add an energy balance, though they should be the same formulation.

andrewlee94 commented:
@bpaul4 There is a way to turn on "symbolic_solver_labels" which will give your the name of the constraint when a failure occurs. I think you can provide this as an argument to the solver call: `solver.solve(model, symbolic_solver_labels=True)`

andrewlee94 commented on idaes/models_extra/gas_solid_contactors/unit_models/tests/test_FB0D.py:
```diff
@@ -345,31 +365,49 @@ def test_scaling(self, iron_oc):
         for t, c in FB0D.solids_energy_holdup_constraints.items():
             assert pytest.approx(
-                1.27323E-9, abs=1e-11
+                1.27324E-7, abs=1e-9
```
This is a very loose tolerance to be checking too - generally you should check to near solver tolerance, thus I tend to use `rel=1e-5`.

bpaul4 commented:
Thank you @andrewlee94 . I implemented the flag above and found that surprisingly it IS the rate constant equation that is failing to evaluate due to a math overflow error `exp(987.723)` which would yield `9.18E+428`. I reformulated the equation in the reaction package to scale down the exponential argument.

chineduobiora commented on idaes/models_extra/gas_solid_contactors/properties/methane_iron_OC_reduction/hetero_reactions.py:
```diff
@@ -456,8 +456,9 @@ def rate_constant_eqn(b, j):
                             * self.solid_state_ref.temperature
                         )
-                    )
+                    ) * 1E-3)
```
Manually scaling the constraints isn't needed. For the failing  test, which is "test_initialize_unscaled" (I think), I suggest reducing the time_set to 360s and nfe to 10 in the iron_oc_unscaled method. Ideally, at the given initial conditions, and time horizon of 3600 you will want to apply a scaling transformation before solving this model.

chineduobiora commented on idaes/models_extra/gas_solid_contactors/unit_models/tests/test_FB0D.py:
```diff
@@ -176,21 +178,231 @@ def test_build(self, iron_oc):
+    def iron_oc_unscaled(self):
+        m = ConcreteModel()
+        m.fs = FlowsheetBlock(
+            default={"dynamic": True, "time_set": [0, 3600], "time_units": pyunits.s}
+        )
+
+        m.fs.gas_props = GasPhaseParameterBlock()
+        m.fs.solid_props = SolidPhaseParameterBlock()
+        m.fs.solid_rxns = HeteroReactionParameterBlock(
+            default={
+                "solid_property_package": m.fs.solid_props,
+                "gas_property_package": m.fs.gas_props,
+            }
+        )
+
+        m.fs.unit = FixedBed0D(
+            default={
+                "gas_property_package": m.fs.gas_props,
+                "solid_property_package": m.fs.solid_props,
+                "reaction_package": m.fs.solid_rxns,
+            }
+        )
+
+        # Discretize time domain
+        m.discretizer = TransformationFactory("dae.finite_difference")
+        m.discretizer.apply_to(m, nfe=100, wrt=m.fs.time, scheme="BACKWARD")
+
```
I suggest reducing the time_set to 360s and nfe to 10 in the iron_oc_unscaled method. This will resolve the issue with the failing "test_initialize_unscaled" test. Ideally, at the given initial conditions, and time horizon of 3600 you will want to apply a scaling transformation before solving this model.

bpaul4 commented:
Thanks @chineduobiora, that resolved the issue.

bpaul4 commented:
Thanks @chineduobiora, that resolved the issue.

chineduobiora reviewed (approved):
Looks good to me.

bpaul4 commented:
@andrewlee94 this is ready for your re-review, please let me know if you have further suggestions before merging this.

andrewlee94 commented on idaes/models_extra/gas_solid_contactors/unit_models/tests/test_FB0D.py:
```diff
@@ -233,13 +233,9 @@ def iron_oc_unscaled(self):
     @pytest.mark.skipif(solver is None, reason="Solver not available")
     @pytest.mark.component
```
This looks like it should be an integration test - it looks like you are solving a (simple) dynamic problem, which is generally more than we put in `component` tests. Note that component tests should take no longer than 10 seconds to run (with very few exceptions).

andrewlee94 reviewed (approved):
I had one comment, but I don't think it should hold up this PR for now.
